hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsRegistryImpl.java

```diff
@@ -16,351 +16,956 @@
+     * Maximum number of times a method annotated with {@link BeforeCollectionCycle}
+     * may throw an exception during collection before it is auto-disabled.
      */
-    public MetricsRegistryImpl(String name, ILogger logger, ProbeLevel minimumLevel) {
-        this.logger = checkNotNull(logger, "logger can't be null");
-        this.minimumLevel = checkNotNull(minimumLevel, "minimumLevel can't be null");
+    private static final int DISABLE_THROWING_UPDATE_LIMIT = 5;
 
-        scheduledExecutorService = new ScheduledThreadPoolExecutor(2,
-                new ThreadFactoryImpl(createThreadPoolName(name, "MetricsRegistry")));
+    private final ConcurrentMap<Class<?>, ProbeAnnotatedType> metaDataCache =
+            new ConcurrentReferenceHashMap<Class<?>, ProbeAnnotatedType>();
+
+    private final ConcurrentMap<Integer, Object> roots =
+            new ConcurrentHashMap<Integer, Object>();
+
+    private final ConcurrentMap<CharSequence, LongGauge> longGauges =
+            new ConcurrentSkipListMap<CharSequence, LongGauge>(SAME_SEQUENCE);
+
+    private final ConcurrentMap<CharSequence, DoubleGauge> doubleGauges =
+            new ConcurrentSkipListMap<CharSequence, DoubleGauge>(SAME_SEQUENCE);
+
+    static ProbeAnnotatedType register(ConcurrentMap<Class<?>, ProbeAnnotatedType> cache, ProbeAnnotatedType typeMetaData) {
+        ProbeAnnotatedType existing = cache.putIfAbsent(typeMetaData.type, typeMetaData);
+        return existing == null ? typeMetaData : existing;
+    }
+
+    static ProbeAnnotatedType getOrCreate(ConcurrentMap<Class<?>, ProbeAnnotatedType> cache, Class<?> type) {
+        ProbeAnnotatedType typeMetaData = cache.get(type);
+        // main goal was to avoid creating expensive metadata but at this point we have to
+        return typeMetaData != null ? typeMetaData : register(cache, new ProbeAnnotatedType(type));
+    }
 
-        if (logger.isFinestEnabled()) {
-            logger.finest("MetricsRegistry minimumLevel:" + minimumLevel);
+    @Override
+    public LongGauge newLongGauge(String name) {
+        LongGauge gauge = longGauges.get(name);
+        if (gauge != null) {
+            return gauge;
         }
+        LongGauge unconnected = Gauges.longGauge();
+        gauge = longGauges.putIfAbsent(name, unconnected);
+        return gauge == null ? unconnected : gauge;
     }
 
     @Override
-    public ProbeLevel minimumLevel() {
-        return minimumLevel;
+    public DoubleGauge newDoubleGauge(String name) {
+        DoubleGauge gauge = doubleGauges.get(name);
+        if (gauge != null) {
+            return gauge;
+        }
+        DoubleGauge unconnected = Gauges.doubleGauge();
+        gauge = doubleGauges.putIfAbsent(name, unconnected);
+        return gauge == null ? unconnected : gauge;
     }
 
-    long modCount() {
-        return sortedProbeInstancesRef.get().mod;
+    @Override
+    public void register(final Object root) {
+        if (root == null) {
+            return;
+        }
```

Comment:
As `register` is the only method called by the main thread we could wrap below part of the method in a try-catch and ignore failing registrations just to be super sure. Although this is tested over and over by integration tests.